url4192078a2f_unsegnor_hipotetizador_tgz-pJ8-hipotetizador_MedirTiemposJ8/src/hipotetizador/Hipotetizador.java

```diff
@@ -0,0 +1,643 @@
+         * {true, true, true}, {false, false, true}, {false, true, false},
+         * {true, true, true}, {false, false, true}, {false, true, false}
+        };
+         */
+
+
+
+        boolean[] muestra;
+
+        for (int i = 0; i < historia.length; i++) {
+            //Generar muestra
+            muestra = historia[i];
+
+            //Imprimir muestra
+            D.d(i + "-" + Muestras.imprimir_muestra(muestra));
+
+            //Introducir muestra
+            h.muestrear(muestra);
+        }
+
+
+
+    }
+
+    private static void historia_inmediata(Hipo h) {
+        boolean[][] historia = {
+            {false},
+            {false},
+            {true},
+            {false},
+            {false},
+            {true},
+            {false},
+            {false},
+            {true},
+            {false},
+            {false},
+            {true}
+        };
+
+        int tventana = 2;
+        float umbral_de_hipotesis = 0.2f;
+        float umbral_de_certeza = 1f;
+        float umbral_de_explicabilidad = 0.2f;
+        boolean generar_subreglas = true;
+
+        h.sinAmbiguedad(historia[0].length, tventana, historia, umbral_de_hipotesis, umbral_de_certeza, umbral_de_explicabilidad, generar_subreglas);
+    }
+
+    private static void serie_numerica(Hipo h) throws IOException {
+
+        InputStreamReader isr = new InputStreamReader(System.in);
+        BufferedReader bf = new BufferedReader(isr);
+
+        //Obtener input en una sola línea separado por espacios
+        System.out.println("Introducir serie numérica en una sola línea separada por espacios");
+        String linea = bf.readLine();
+
+        //Separar la entrada por espacios
+        String[] s_num = linea.split(" ");
+
+        //Convertir los números en binario
+        String[] s_bin = new String[s_num.length];
+        //Almacenamos la longitud máxima
+        int maxlength = 0;
+        for (int i = 0; i < s_num.length; i++) {
+            s_bin[i] = Integer.toBinaryString(Integer.parseInt(s_num[i]));
+            maxlength = Math.max(maxlength, s_bin[i].length());
+        }
+
+        //La longitud máxima es el número de entradas
+        int nentradas = maxlength;
+        //La cantidad de números es el tamaño de la historia
+        int t_historia = s_bin.length;
+
+        //Componemos la historia
+        boolean[][] historia = new boolean[t_historia][nentradas];
+        //Para cada línea de la historia
+        for (int i = 0; i < t_historia; i++) {
+            //Leemos la cadena binaria correspondiente
+            String c_bin = s_bin[i];
+            int l = c_bin.length();
+            //Vamos rellenando la historia hasta donde lleguemos
+            for (int j = 0; j < l; j++) {
+                historia[i][j] = c_bin.charAt((l-1)-j) == '0' ? false : true;
+            }
+            //Después llenamos el resto de ceros
+            for (int j = l; j < nentradas; j++) {
+                historia[i][j] = false;
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 237:249, 464:476, 591:603.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.